### PR TITLE
[baremetal] Add SCA countermeasures to SHA-256 module

### PR DESCRIPTION
**Pull request by jarlamsa**
This PR adds SCA countermeasures to SHA-256 module on places where
get/put_uint32 on secret data is used.

When reading the input, the buffer will be initialised with random data
and the reading will start from a random offset. When writing the data,
the output will be initialised with random data and the writing will
start from a random offset.

## Status
**READY**

### CODE REVIEW

File: library/sha256.c [hunk 1]
```diff
@@ -367,17 +382,33 @@ int mbedtls_sha256_finish_ret( mbedtls_sha256_context *ctx,
...
     {
-        (void)mbedtls_platform_put_uint32_be( &output[o_pos],
-                                              ctx->state[s_pos] );
-    }
+        uint32_t initval = mbedtls_platform_random_in_range(0xffffffff);
+        uint32_t offset = mbedtls_platform_random_in_range(7);
+
+        for( o_pos = 0; o_pos < 32; o_pos += 4 )
+        {
+            (void)mbedtls_platform_put_uint32_be( &output[o_pos], initval );
```
Comment (artokin) [inline library/sha256.c]: Do we need to use `mbedtls_platform_put_uint32_be` here to set initval?
Could we use just `memset` or even Mbed TLS memset?

File: library/sha256.c [hunk 2]
```diff
@@ -367,17 +382,33 @@ int mbedtls_sha256_finish_ret( mbedtls_sha256_context *ctx,
      * Output final state
      */
 
-    for( s_pos = 0, o_pos = 0; s_pos < 7; s_pos++, o_pos += 4 )
     {
-        (void)mbedtls_platform_put_uint32_be( &output[o_pos],
-                                              ctx->state[s_pos] );
-    }
+        uint32_t initval = mbedtls_platform_random_in_range(0xffffffff);
```
Comment (artokin) [inline library/sha256.c]: very minor: could we avoid using code block

File: library/sha256.c [hunk 3]
```diff
@@ -196,11 +196,26 @@ int mbedtls_internal_sha256_process( mbedtls_sha256_context *ctx,
...
 
 #if defined(MBEDTLS_SHA256_SMALLER)
-    for( i = 0; i < 64; i++ )
     {
-        if( i < 16 )
+        uint32_t initval = mbedtls_platform_random_in_range(0xffffffff);
+        uint32_t offset = mbedtls_platform_random_in_range(16);
+        for( i = 0; i < 16; i++ )
+        {
+            W[i] = initval;
```
Comment (artokin) [inline library/sha256.c]: Could we use `memset` or Mbed TLS version of it?

File: library/sha256.c (hunk 1 above)
Comment (jarlamsa) [reply library/sha256.c]: Done

File: library/sha256.c (hunk 3 above)
Comment (jarlamsa) [reply library/sha256.c]: Done

**Comment by jarlamsa** [comment]
@artokin thank you for your review comments, I think I have addressed those, please re-review.

**Comment by jarlamsa** [comment]
2 approving reviews, passing CI -> ready for merge.

File: library/sha256.c [hunk 4]
```diff
@@ -196,11 +196,23 @@ int mbedtls_internal_sha256_process( mbedtls_sha256_context *ctx,
         A[i] = ctx->state[i];
 
 #if defined(MBEDTLS_SHA256_SMALLER)
-    for( i = 0; i < 64; i++ )
     {
```
[the remark this answers is not in the document]
Comment (sbutcher-arm) [reply library/sha256.c]: That's fine. Leave it as is then.

File: library/sha256.c [hunk 5]
```diff
@@ -196,11 +196,23 @@ int mbedtls_internal_sha256_process( mbedtls_sha256_context *ctx,
...
-        if( i < 16 )
+        uint32_t offset = mbedtls_platform_random_in_range(16);
+        mbedtls_platform_memset( W, 0, sizeof( W ) );
+
+        for( i = offset; i < 16; i++ )
```
[the remark this answers is not in the document]
Comment (jarlamsa) [reply library/sha256.c]: Might be, that would need to span then to every other place we are having similar approach. Definitely nice to have if this is going to be upstreamed.

**Comment by jarlamsa** [comment]
@sbutcher-arm I don't think there are any comments left that need addressing, if so, please re-review.

**Pull request merged**